ably/proto_message_decoding_test.go

```diff
@@ -0,0 +1,147 @@
+package ably
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/json"
+	"os"
+	"strings"
+	"testing"
+
+	"github.com/ably/ably-go/ably/internal/ablyutil"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+type fixture struct {
+	Data          string `json:"data"`
+	Encoding      string `json:"encoding"`
+	ExpectedType  string `json:"expectedType"`
+	ExpectedValue any    `json:"expectedValue"`
+}
+
+func loadFixtures() ([]fixture, error) {
+	var dec struct {
+		Messages []fixture
+	}
+
+	// We can't embed the fixtures as go:embed forbids embedding of resources above the current directory.
+	text, err := os.ReadFile("../common/test-resources/messages-encoding.json")
+	if err != nil {
+		return nil, err
+	}
+
+	err = json.Unmarshal(text, &dec)
+	return dec.Messages, err
+}
+
+func Test_decodeMessage(t *testing.T) {
+	fixtures, err := loadFixtures()
+	require.NoError(t, err, "failed to load test fixtures")
+	for _, f := range fixtures {
+		t.Run(f.Data, func(t *testing.T) {
+			msg := Message{
+				Data:     f.Data,
+				Encoding: f.Encoding,
+			}
+			decodedMsg, err := msg.withDecodedData(nil)
+			assert.NoError(t, err)
+			switch f.ExpectedType {
+			case "string":
+				assert.IsType(t, "string", decodedMsg.Data)
+				assert.Equal(t, f.ExpectedValue, decodedMsg.Data)
+			case "jsonObject":
+				assert.IsType(t, map[string]any{}, decodedMsg.Data)
+				assert.Equal(t, f.ExpectedValue, decodedMsg.Data)
+
+			case "jsonArray":
+				assert.IsType(t, []any{}, decodedMsg.Data)
+				assert.Equal(t, f.ExpectedValue, decodedMsg.Data)
+			case "binary":
+				assert.IsType(t, []byte{}, decodedMsg.Data)
+				if f.ExpectedValue == nil {
```

Comment:
Should all fixtures not have an expected value?

Reply:
if they have a legal decoding, yes
But "3q2+7w==" (which is base64 for 0xdeadbeef) does not have an ExpectedValue in messages-encoding.json
(though it does have an expectedHexValue).

Reply:
So should this be checking `expectedHexValue`?

Reply:
done in https://github.com/ably/ably-go/pull/588/commits/707c94f5163d8d1422cc6efbd0a11e6b102094d5